java/connector-node/risingwave-connector-service/src/main/java/com/risingwave/metrics/ConnectorNodeMetrics.java

```diff
@@ -167,4 +197,48 @@ public static void incErrorCount(String sinkType, String ip) {
     public static void setRamUsage(String ip, long usedRamInBytes) {
         ramUsage.labels(ip).set(usedRamInBytes);
     }
+
+    public static void recordCdcEngineFailure(
+            String sourceType,
+            String sourceId,
+            String connectorClass,
+            String message,
+            Throwable error) {
+        String kind = classifyCdcEngineFailureKind(message, error);
+        cdcEngineFailureTotal.labels(sourceType, sourceId, connectorClass, kind).inc();
+    }
+
+    private static String classifyCdcEngineFailureKind(String message, Throwable error) {
+        // Use the actual exception type as error_kind, instead of keyword matching.
+        // This stays low-cardinality in practice and is much more precise for alerting.
+        Throwable root = rootCause(error);
+        if (root != null) {
+            String simple = root.getClass().getSimpleName();
+            if (simple != null && !simple.isBlank()) {
+                return simple;
+            }
+            // Fallback to the full class name if needed (still finite), but keep it short.
+            String full = root.getClass().getName();
+            return full == null || full.isBlank() ? "Unknown" : full;
+        }
+        // If we don't have a Throwable at all, we can't reliably infer the type.
+        // Keep a stable fallback value to avoid label explosion.
+        return "Unknown";
+    }
+
+    private static Throwable rootCause(Throwable t) {
+        if (t == null) {
+            return null;
+        }
+        Throwable cur = t;
+        // Guard against cause cycles or overly deep chains.
```

Comment:
Can we add a TODO, I think there can be some de-duplication done in the future or cycle detection done. E.g. we can filter out useless causes